bin/fetch-from-genbank

```diff
@@ -1,40 +1,72 @@
-from sys import argv, exit, stdin, stdout, stderr
+"""
+Download all SARS-CoV-2 sequences and their curated metadata from GenBank via
+NCBI Virus.
 
-Entrez.email = environ.get("NCBI_EMAIL", "hello@nextstrain.org")
+Outputs newline-delimited JSON records.
 
-if not argv[1:]:
-    print(f"Reading metadata.tsv from stdin", file = stderr)
-    metadata_file = stdin
+The request this program makes is based on observing the network activity that
 
-elif len(argv[1:]) == 1:
-    print(f"Reading metadata.tsv from {argv[1]!r}", file = stderr)
-    metadata_file = argv[1]
+    https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus?SeqType_s=Nucleotide&VirusLineage_ss=Severe%20acute%20respiratory%20syndrome%20coronavirus%202,%20taxid:2697049
 
-else:
-    print(f"usage: {__file__} [<metadata.tsv>]", file = stderr)
-    exit(1)
+performs after clicking through the download interface.  Some tweaks were made
+by comparing different download requests and guessing, which allows us to
+download the metadata + sequence in the same request instead of two.
+"""
+import csv
+import json
+import requests
+from sys import stdout
 
-metadata = pd.read_table(metadata_file)
+endpoint = "https://www.ncbi.nlm.nih.gov/genomes/VirusVariation/vvsearch2/"
+params = {
+    # Search criteria
+    'fq': [
+        '{!tag=SeqType_s}SeqType_s:("Nucleotide")', # Nucleotide sequences (as opposed to protein)
+        'VirusLineageId_ss:(2697049)',              # NCBI Taxon id for SARS-CoV-2
+    ],
 
-accessions = list(
-    metadata["genbank_accession"]
-        .replace("?", pd.NA)
-        .dropna()
-        .drop_duplicates()
-)
+    # Unclear, but seems necessary.
+    'q': '*:*',
 
-print(f"Fetching {len(accessions):,} accessions from GenBank", file = stderr)
+    # Response format
+    'cmd': 'download',
+    'dlfmt': 'csv',
+    'fl': ','.join(
+        ':'.join(names) for names in [
+            # Pairs of (output column name, source data field).  These are pulled
+            # from watching requests from the UI.
+            #
+            # XXX TODO: Is the full set source data fields documented
+            # somewhere?  Is there more info we could be pulling that'd be
+            # useful?
+            #   -trs, 13 May 2020
+            ('genbank_accession',       'id'),
+            ('database',                'SourceDB_s'),
+            ('title',                   'Definition_s'),
+            ('location',                'CountryFull_s'),
+            ('collected',               'CollectionDate_s'),
+            ('submitted',               'CreateDate_dt'),
+            ('length',                  'SLen_i'),
+            ('host',                    'Host_s'),
+            ('isolation_source',        'Isolation_csv'),
+            ('biosample_accession',     'BioSample_s'),
+            ('authors',                 'Authors_csv'),
+            ('publications',            'PubMed_csv'),
+            ('sequence',                'Nucleotide_seq'),
```

Comment:
I found some additional fields that would be useful while playing around with the API on Postman:
```
('region', 'Region_s`),
('strain', 'Isolate_s')
```

Reply:
Awesome! :raised_hands: I've added these in another commit.